app/src/main/java/pm/gnosis/heimdall/ui/safe/pairing/PairingAuthenticator.kt

```diff
@@ -0,0 +1,193 @@
+        ) : ViewUpdate()
+
+        data class Pairing(
+            val isLoading: Boolean,
+            val pairingResult: PairingResult? = null
+        ) : ViewUpdate()
+
+        data class Authenticator(
+            val info: AuthenticatorSetupInfo
+        ) : ViewUpdate()
+    }
+
+}
+
+class PairingAuthenticatorViewModel @Inject constructor(
+    private val recoverSafeOwnersHelper: RecoverSafeOwnersHelper,
+    private val gnosisSafeRepository: GnosisSafeRepository,
+    private val accountsRepository: AccountsRepository,
+    private val tokenRepository: TokenRepository,
+    private val transactionExecutionRepository: TransactionExecutionRepository,
+    private val pushServiceRepository: PushServiceRepository,
+    private val moshi: Moshi,
+    private val appDispatcher: ApplicationModule.AppCoroutineDispatchers
+) : PairingAuthenticatorContract() {
+
+    private val errorHandler = CoroutineExceptionHandler { _, e ->
+        viewModelScope.launch {
+            Timber.e(e)
+        }
+    }
+
+    override val observableState: LiveData<ViewUpdate>
+        get() = _state
+    private val _state = MutableLiveData<ViewUpdate>()
+
+    private lateinit var safeAddress: Solidity.Address
+    private lateinit var safeInfo: SafeInfo
+
+    override fun setup(safeAddress: Solidity.Address) {
+        this.safeAddress = safeAddress
+    }
+
+    override fun estimate() {
+
+        viewModelScope.launch(appDispatcher.background + errorHandler) {
+
+            val paymentToken = tokenRepository.loadPaymentToken(safeAddress).await()
+
+            safeInfo = gnosisSafeRepository.loadInfo(safeAddress).awaitFirst()
+
+            val owner = accountsRepository.signingOwner(safeAddress).await()//safeInfo.owners[0]
+            val extension = Solidity.Address(BigInteger.valueOf(Long.MAX_VALUE))
+            val pairingTransaction = recoverSafeOwnersHelper.buildRecoverTransaction(
+                safeInfo,
+                safeInfo.owners.subList(2, safeInfo.owners.size).toSet(),
+                setOf(owner.address, extension)
+            )
+
+            val safeOwner = accountsRepository.signingOwner(safeAddress).await()
+
+            val executeInfo = transactionExecutionRepository.loadExecuteInformation(safeAddress, paymentToken.address, pairingTransaction, safeOwner).await()
+            val gasFee = executeInfo.gasCosts()
+
+            _state.postValue(
+                ViewUpdate.Balance(
+                    paymentToken.displayString(executeInfo.balance),
+                    ERC20TokenWithBalance(paymentToken, gasFee).displayString(roundingMode = RoundingMode.UP),
+                    paymentToken.displayString(executeInfo.balance - gasFee),
+                    paymentToken.symbol,
+                    executeInfo.balance > gasFee
+
+                )
+            )
+        }
+    }
+
+
+    override fun loadAuthenticatorInfo() {
+
+        viewModelScope.launch(appDispatcher.background + errorHandler) {
```

Comment:
Should we have some kind of loading spinner?

Reply:
there shouldn't be any noticeable delay for the user